interfaces/acados_template/acados_template/mpc_utils.py

```diff
@@ -28,73 +28,455 @@
+
+    def update_all(self, acados_solver: AcadosOcpSolver):
+        """
+        Update the parameter values and global parameter values from the acados solver.
+        ATTENTION: Currently only parameter values are updated. Reference values, bounds, etc. are not updated.
+        """
+        N = self.ocp.dims.N
+        if self.ocp.dims.np > 0:
+            new_parameter_values = np.zeros((self.ocp.dims.np, N))
+            for i in range(N):
+                new_parameter_values[:, i] = acados_solver.get(i, 'p')
+            self.__parameter_values = new_parameter_values
+
+        if acados_solver.save_p_global is False:
+            print('\nCan not set \'p_global\', since the solver does not store these values by default. '
+                  'In order to update \'p_global\', please set the option \'save_p_global=True\' in the '
+                  'instantiation of the acados solver.')
+        else:
+            self.__p_global_values = acados_solver.get_flat('p_global')
+
+    def _set_check_parameters(
+            self, p: np.ndarray = None,
+            p_global: np.ndarray = None):
+        if p is not None:
+            if p.shape == self.__parameter_values:
+                self.__parameter_values = p
+            else:
+                raise ValueError(
+                    f"Parameter vector 'p' has wrong shape {p.shape} instead of {self.__parameter_values.shape}.")
+
+        if p_global is not None:
+            if p_global.shape == self.__p_global_values:
+                self.__p_global_values = p_global
+            else:
+                raise ValueError(
+                    f"Global parameter vector 'p_global' has wrong shape {p_global.shape} instead of {self.__p_global_values.shape}.")
+
+    def evaluate(self, x: np.ndarray,
+                 u: np.ndarray,
+                 step: int = 0,
+                 p: np.ndarray = None,
+                 p_global: np.ndarray = None,
+                 ) -> dict:
+        """
+        Evaluates the cost and constraints at a given stage of the OCP. For a closed-loop evaluation the stage is
+        typically 0. If parameter values are provided, they are also set in the evaluator.
+        @param x: state vector
+        @param u: control input vector
+        @param step: stage index (0 <= stage < N)
+        @param p: parameter vector used for evaluation (optional)
+        @param p_global: global parameter vector used for evaluation (optional)
+
+        @return: dictionary with the following keys:
+            - 'cost': total cost of the transition
+            - 'cost_without_slacks': total cost of transition without slack penalties
+            - 'cost_slacks': total cost of slack penalties
+            - 'violation_soft_constraints': individual violation of soft constraints (equal to slacks)
+            - 'violation_hard_constraints': individual violation of hard constraints
+        """
+        self._set_check_parameters(p, p_global)
```

Comment:
not sure if we also want to store the parameters, changes the _state_ of the evaluator..

Reply:
questionable but agree